### ISSUE TITLE
Simple recursive function may create an infinite loop without explicit return statement

### ISSUE THREAD

**Issue by charlie-rudenstal** [babel/babel#822]
This recursive function 

```
function getConfigKey(config, key) {
	if (config[key]) return config[key];
	var parts = key.split('.');
	var firstPart = parts.shift();
	if(config[firstPart]) {
		return getConfigKey(config[firstPart], parts.join('.'));
	}
}
```

Transpiles to 

````
function getConfigKey(_x, _x2) {
	_function: while (true) {
		var config = _x,
		    key = _x2;
		parts = firstPart = undefined;
		if (config[key]) {
			return config[key];
		}var parts = key.split(".");
		var firstPart = parts.shift();
		if (config[firstPart]) {
			_x = config[firstPart];
			_x2 = parts.join(".");
			continue _function;
		}
	}
}
```

The transpiled version will create an infinite loop for keys that are not found. This problem did not exist before the transpilation.

The problem can be solved by adding an explicit return statement (which might be good anyway, since it makes the code more readable)

```
function getConfigKey(config, key) {
	if (config[key]) return config[key];
	var parts = key.split('.');
	var firstPart = parts.shift();
	if(config[firstPart]) {
		return getConfigKey(config[firstPart], parts.join('.'));
	} else {
	  return; // explicit return because babel.js creates an infinite while(true) loop otherwise
	}	  
}
```

But it might still be a bug to consider, since it is an unexpected behavior. Took me a while to track down after transpiling our existing codebase .

**Comment by sebmck** [comment]
@RReverser Any ideas?

**Comment by RReverser** [comment]
Please let me know if attached patch doesn't work.

**Comment by sebmck** [comment]
Fixed as of 4.4.0, thanks!
